proposals/0066-feature-gate-threshold-automation.md

```diff
@@ -0,0 +1,245 @@
+activation.
+
+This SIMD proposes 3 steps over the span of one epoch to determine which
+feature gates are queued for activation and assess stake support:
+
+1. On an epoch boundary, the runtime generates a list of feature-gates queued
+for activation.
+2. Validators signal which of the queued feature-gates they support.
+3. On the next epoch boundary, the runtime activates the feature-gates that
+have the necessary stake support
+
+### Representing the Current Feature Set
+
+As mentioned above, a BPF program is being deployed that will replace the
+non-existent account `Feature111111111111111111111111111111111111`, which is
+the owner of all feature accounts.
+
+The Feature Gate program will utilize two PDAs to track the authoritative
+feature set queued for activation. One PDA (the Feature-Request PDA) will
+store newly requested feature activations and the other (the Feature-Queue
+PDA) will store the previous epoch’s requested activations.
+
+When a new epoch begins, the Feature-Request PDA will be an empty list. When
+key-holders submit the CLI command to activate a feature, the Feature Gate
+Program appends the feature ID to the Feature-Request set at the same time as
+creating the feature account . Revoked features are removed from the
+Feature-Request list. At the end of the epoch, the Feature-Request set is
+written to the feature queue PDA, where it becomes immutable. The
+Feature-Request PDA is then reset to an empty list again.
+
+The newly created immutable Feature-Queue can then be used by nodes to signal
+support for potential activation in the next epoch.
+
+Note that for any given epoch, the epoch will begin with an immutable set of
+previously requested feature activations, and end with the activation of those
+with enough support. This process takes one epoch. Therefore, it will take at
+least one more epoch than it currently does from the time a key-holder submits
+a feature for activation via CLI to when it is actually activated.
+
+### Signaling Support for Features
+
+With an on-chain reference point to determine the features queued for
+activation, nodes can signal support for specific features in the queue.
+
+A node signals its support for a feature set by sending a transaction to the
+Feature Gate program, signed by their authorized voter, to store a record of
+their supported feature set.
+This supported feature set would be a bit mask of the list of queued features,
+and it would be stored in a PDA derived from their vote address.
+Nodes simply examine the Feature-Queue list, and mark a 1 for any features they
+have in their feature set. The rest are marked 0.
+
+In the previous example, a node running v1.14.25 would signal the following bit
+mask:
+
+```
+1   0   1   0 
+```
+
+Similarly, a node running v1.16.0 would signal:
+
+```
+1   1   1   1 
+```
+
+Validators send these transactions signaling their support for a queued set of
+features on the first slot of every epoch and on startup after any reboot.
```

Comment:
Rather than sending the transaction on the first slot of every epoch (thundering herd!), how about they start attempting to send the transaction at a random time during the first 25% of the epoch and then retry randomly every N slots. 

At restart they send immediately once the health check passes, and then also retry randomly every N slots

Reply:
Yeah this would work too. As long as they send their initial signal early enough to account for any necessary retries it's ok. Could even be 50% if we wanted.